lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp

```diff
@@ -0,0 +1,155 @@
+//===-- OperatingSystemSwiftTasks.cpp -------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#if LLDB_ENABLE_SWIFT
+
+#include "OperatingSystemSwiftTasks.h"
+#include "Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.h"
+#include "Plugins/Process/Utility/ThreadMemory.h"
+#include "lldb/Core/Debugger.h"
+#include "lldb/Core/Module.h"
+#include "lldb/Core/PluginManager.h"
+#include "lldb/Target/Process.h"
+#include "lldb/Target/Thread.h"
+#include "lldb/Target/ThreadList.h"
+#include "lldb/Utility/LLDBLog.h"
+
+#include <memory>
+
+using namespace lldb;
+using namespace lldb_private;
+
+LLDB_PLUGIN_DEFINE(OperatingSystemSwiftTasks)
+
+void OperatingSystemSwiftTasks::Initialize() {
+  PluginManager::RegisterPlugin(GetPluginNameStatic(),
+                                GetPluginDescriptionStatic(), CreateInstance,
+                                nullptr);
+}
+
+void OperatingSystemSwiftTasks::Terminate() {
+  PluginManager::UnregisterPlugin(CreateInstance);
+}
+
+OperatingSystem *OperatingSystemSwiftTasks::CreateInstance(Process *process,
+                                                           bool force) {
+  if (!process || !process->GetTarget().GetSwiftUseTasksPlugin())
+    return nullptr;
+
+  Log *log = GetLog(LLDBLog::OS);
+  std::optional<uint32_t> concurrency_version =
+      SwiftLanguageRuntime::FindConcurrencyDebugVersion(*process);
+  if (!concurrency_version) {
+    LLDB_LOG(log,
+             "OperatingSystemSwiftTasks: did not find concurrency module.");
+    return nullptr;
+  }
+
+  if (*concurrency_version <= 1)
+    return new OperatingSystemSwiftTasks(*process);
+  LLDB_LOGF(log,
```

Comment:
Did you intend to only log the version in the failure case?

Reply:
Also, if you've encountered a more recent version of the concurrency library that this lldb recognizes, we should tell the user that.  Could there be a `Status` path threaded to  this `Initialize`.

Reply:
Yes, I was originally only logging the problematic case, but there is not harm in logging it all the time, so I'll move the log statement before the conditional

Reply:
>Also, if you've encountered a more recent version of the concurrency library that this lldb recognizes, we should tell the user that. Could there be a Status path threaded to this Initialize.

Is this how we do user warnings? I was grepping around and found `Debugger::ReportWarning`, which a matching `std::once_flag` flag. If this is the standard way of doing things, and to make sure a warning for this plugin is only reported once, it seems like the implementation would be to call ReportWarning inside `OperatingSystemSwiftTasks` directly, and have that class own the std::once flag. (Note that the "CreateInstance" methods are called many times, e.g. every time libraries are loaded)

Reply:
I wasn't reading this right.  There really isn't anything that can fail here, since we're just looking for a module by name, and then the version will only be an empty optional if we didn't find the module.  And not finding the module is fine, we just might not be using swift concurrency...  So there's no warning to issue here, IIUC.

Reply:
What you've added here looks right.  We only warn if we found the concurrence library but it looks funny to us.